### DISCUSSION TITLE
combinations of two particles along ID dimension

### DISCUSSION THREAD

**Discussion by miniufo** [pydata/xarray#8121] [Q&A]
I have a Lagrangian particle dataset:
```python
<xarray.Dataset>
Dimensions:  (time: 2000, ID: 2000)
Coordinates:
  * time     (time) float32 0.0 0.01 0.02 0.03 0.04 ... 19.96 19.97 19.98 19.99
  * ID       (ID) float32 0.0 1.0 2.0 3.0 ... 1.997e+03 1.998e+03 1.999e+03
Data variables:
    xpos     (ID, time) float32 ...
    ypos     (ID, time) float32 ...
    uvel     (ID, time) float32 ...
    vvel     (ID, time) float32 ...
```

which contains 2000 Lagrangian particles (`ID`).  I need to perform relative dispersion calculation in which every two particles should be grouped into a pair of particles.  For example, a pair of particles of IDs 0 and 1 (IDs for a single pair should not be the same).  I can do this using:
```python
for j in range(ds.ID):
    for i in range(j, len(ds.ID)):
        p1 = ds.sel(ID=j)
        p2 = ds.sel(ID=i)
        stat = some_function(p1, p2)
```

But number of pairs (2000*1999/2) is quite large so that the explicit loop is very slow.  Just wondering if there is more efficient and elegant way of doing this?  Or can I convert the dataset from (`time`, `ID`) into (`time`, `pair`, `ID`) so that the `pair` dimension is all the combinations of two ID (which is convenient for iteration over all pairs)?

It is the best the result `stat` is also of dimensions (`time` `pair`).  But if `pair` is too large, one can do reduction like `mean` to get average over `pair` dimension.

**Comment by keewis** [comment]
if you have a list of indexes, you can use `isel` for this:
```python
import numpy as np
import itertools

indices = xr.DataArray(np.array(list(itertools.combinations(range(2000), 2))), dims=("ID", "pair"))
reordered = ds.isel(ID=indices)
```
of course, this will create an additional dimension of size 1999000, so you might have to use `dask` to chunk the dataset first if it doesn't fit into memory (use `ds.chunk({"time": chunksize})` where `chunksize` depends on the size of your . It might also be worth vectorizing `some_function` such that it can handle computing the stats for a large amount of pairs at once (which usually makes the whole operation *a lot* faster):
```python
stats = some_function(reordered.isel(pair=0), reordered.isel(pair=1))
# or, even better:
stats = reordered.reduce(some_function, dims="pair")
stats = some_function(reordered, dims="pair")
```
The last version can be implemented using `apply_ufunc`, but would require a low-level function that consumes and produces `numpy`-like arrays.

**Reply by miniufo** [reply]
That great!  I will try re-ordering the dataset.  Is it possible to re-chunk along pair dimension?  I guess the dataset is stored in `time` -continuous fashion, so re-chunk `pair` would be more efficient?
